source/util/move_to_front.h

```diff
@@ -0,0 +1,649 @@
+    assert(node);
+    return nodes_.at(node).size;
+  }
+
+  // Returns mutable reference to height of |node|.
+  uint32_t& MutableHeightOf(uint32_t node) {
+    assert(node);
+    return nodes_.at(node).height;
+  }
+
+  // Returns true iff |node| is left child of its parent.
+  bool IsLeftChild(uint32_t node) const {
+    assert(node);
+    return LeftOf(ParentOf(node)) == node;
+  }
+
+  // Returns true iff |node| is right child of its parent.
+  bool IsRightChild(uint32_t node) const {
+    assert(node);
+    return RightOf(ParentOf(node)) == node;
+  }
+
+  // Returns true iff |node| has no relatives.
+  bool IsOrphan(uint32_t node) const {
+    assert(node);
+    return !ParentOf(node) && !LeftOf(node) && !RightOf(node);
+  }
+
+  // Returns true iff |node| is in the tree.
+  bool IsInTree(uint32_t node) const {
+    assert(node);
+    return node == root_ || !IsOrphan(node);
+  }
+
+  // Returns the height difference between right and left subtrees.
+  int BalanceOf(uint32_t node) const {
+    return int(HeightOf(RightOf(node))) - int(HeightOf(LeftOf(node)));
+  }
+
+  // Updates size and height of the node, assuming that the children have
+  // correct values.
+  void UpdateNode(uint32_t node);
+
+  // Returns the most LeftOf(LeftOf(... descendent which is not leaf.
+  uint32_t LeftestDescendantOf(uint32_t node) const {
+    uint32_t parent = 0;
+    while (node) {
+      parent = node;
+      node = LeftOf(node);
+    }
+    return parent;
+  }
+
+  // Returns the most RightOf(RightOf(... descendent which is not leaf.
+  uint32_t RightestDescendantOf(uint32_t node) const {
+    uint32_t parent = 0;
+    while (node) {
+      parent = node;
+      node = RightOf(node);
+    }
+    return parent;
+  }
+
+  // Inserts node in the tree. The node must be an orphan.
+  void InsertNode(uint32_t node);
+
+  // Removes node from the tree. May change value_to_node_ if removal uses a
+  // scapegoat. Returns the removed (orphaned) handle for recycling. The
+  // returned handle may not be equal to |node| if scapegoat was used.
+  uint32_t RemoveNode(uint32_t node);
+
+  // Rotates |node| left, reassigns all connections and returns the node
+  // which takes place of the |node|.
+  uint32_t RotateLeft(const uint32_t node);
+
+  // Rotates |node| right, reassigns all connections and returns the node
+  // which takes place of the |node|.
+  uint32_t RotateRight(const uint32_t node);
+
+  // Root node handle. The tree is empty if root_ is 0.
+  uint32_t root_ = 0;
+
+  // Incremented counters for next timestamp and value.
+  uint32_t next_timestamp_ = 1;
+
+  // Holds all tree nodes. Indices of this vector are node handles.
+  std::vector<Node> nodes_;
+
+  // Maps ids to node handles.
+  std::unordered_map<Val, uint32_t> value_to_node_;
+};
+
+template <typename Val>
+bool MoveToFront<Val>::Insert(const Val& value) {
+  auto it = value_to_node_.find(value);
+  if (it != value_to_node_.end() && IsInTree(it->second))
```

Comment:
*Might* be worth a comment to say that reinsertion of a node that had been removed will reuse its old node.  That would explain the second clause here.